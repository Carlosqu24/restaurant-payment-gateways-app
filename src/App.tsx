import { Box, Paper, CssBaseline, ThemeProvider } from "@mui/material";
import { createTheme } from "@mui/material/styles";
import { BrowserRouter as Router, Routes, Route } from "react-router-dom";
import { routes as appRoutes } from "./routes";
import Navbar from "./components/Navbar";
import Footer from "./components/Footer";
import NotFound from "./pages/NotFound";

export function App() {

  // define theme
  const theme = createTheme({
    palette: {
      primary: {
        light: "#63b8ff",
        main: "#0989e3",
        dark: "#005db0",
        contrastText: "#000",
      },
      secondary: {
        main: "#4db6ac",
        light: "#82e9de",
        dark: "#00867d",
        contrastText: "#000",
      },
    },
   
    typography: {
      fontFamily: `"Inter", "IBM Plex Sans"`,
      
    },
  });

  return (
    <ThemeProvider theme={theme} >

      <Router>
        <Navbar />

      <div className="p-4" >
      <Routes>
          {appRoutes.map((route) => (
            <Route
              key={route.key}
              path={route.path}
              element={<route.component />}
            />
          ))}
          <Route path="*" element={<NotFound />}></Route>
        </Routes>
      </div>


      </Router>
    </ThemeProvider>
  );
}

export function WrappedApp() {
  return <App />
}
