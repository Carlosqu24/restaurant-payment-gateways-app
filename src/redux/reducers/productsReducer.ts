import { createAsyncThunk, createSlice, current } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'

import { Product } from "../../models/product"
import { getAllProducts } from '../../services/product'

interface ProductsState {
  data: Product[],
  isLoading: boolean,
  errorText: string | null
}

const productsReducerInitialState: ProductsState = {
  data: [],
  isLoading: false,
  errorText: null
}

export const productsReducerTestingState = {
    data: [{
      id: 1,
      name: "Product 1",
      description: "Description 1",
      salePrice: 2,
      rating: 3,
      imageUrl: "",
      category: "Meats"
    }, {
      id: 2,
      name: "Product 2",
      description: "Description 1",
      salePrice: 6,
      rating: 7,
      imageUrl: "",
      category: "Meats"
    }],
    errorText: "",
    isLoading: false
}

export const fetchAllProducts = createAsyncThunk(
  'products/fetchAllProducts',
  async () => {
    const response = await getAllProducts()
    return response
  }
)

export const saveProduct = createAsyncThunk(
  'products/saveProduct',
  async (newProduct) => {

  }
)

export const productsSlice = createSlice({
  name: 'products',
  initialState: productsReducerInitialState,
  reducers: {
    createNewProduct: (state, action) => {
      state.data.push(action.payload)

      // console.log({ payload: action.payload, state: current(state) })
    },
    editProduct: () => {
      // const foundProduct = current(state).data.find(product => product.id === action.payload)
    },
    deleteProduct: () => {
      // const foundProduct = current(state).data.find(product => product.id === action.payload)
    }
  },
  extraReducers: (builder) => {
    builder.addCase(fetchAllProducts.fulfilled, (state: any, action) => {
      // console.log({ payload: action.payload })
      // state.entities.push(action.payload)
      state.data = action.payload

      // state.products.data = action.payload
    })
  }
})

export const { createNewProduct } = productsSlice.actions

export default productsSlice.reducer