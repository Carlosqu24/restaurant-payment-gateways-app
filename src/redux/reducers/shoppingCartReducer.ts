// @ts-nocheck
import { createSlice, current } from '@reduxjs/toolkit'

import { ShoppingCartProduct } from "../../models/product"

// interface PurchaseDetails {
//     subtotal: 0,
//     taxAmount: 0,
//     fees: {
//       shippingFee: 0,
//       processingFee: 0
//     },
//     total: 0
//   }

export interface PurchaseDetails {
    subTotal: number
    taxAmount: number
    total: number
}

interface ShoppingCartState {
    dataList: ShoppingCartProduct[],
    isLoading: boolean,
    errorText: string | null,
    purchaseDetails: PurchaseDetails
}

const shoppingCartInitialState: ShoppingCartState = {
    dataList: [],
    isLoading: false,
    errorText: null,
    purchaseDetails: {
        subTotal: 0,
        taxAmount: 0,
        total: 0,
    }
}

export const shoppingCartSlice = createSlice({
    name: 'shoppingCart',
    initialState: shoppingCartInitialState,
    reducers: {
        addShoppingCartProduct: (state, action) => {

            const newCartProduct: ShoppingCartProduct = {
                ...action.payload,
                quantity: 1,
                totalPrice: 1 * action.payload.salePrice
            }

            state.dataList.push(newCartProduct)


            // Calculate purchaseDetails

            const subTotal = state.dataList.reduce(
                (prev, actual) => prev + (actual.salePrice * actual.quantity),
                0
            )
            const taxAmount = subTotal * 0.13
            const total = subTotal + taxAmount

            state.purchaseDetails.subTotal = subTotal
            state.purchaseDetails.taxAmount = taxAmount
            state.purchaseDetails.total = total
        },
        incrementShoppingCartProductQuantity: (state, action) => {
            const cartProductId = action.payload
            const foundCartProduct = current(state).dataList.find(cartProduct => cartProduct.id === cartProductId)

            const newCartProductQuantity = foundCartProduct?.quantity + 1

            const editedCartProduct: ShoppingCartProduct = {
                ...foundCartProduct,
                quantity: newCartProductQuantity,
                totalPrice: newCartProductQuantity * foundCartProduct?.salePrice
            }

            const filteredData = current(state).dataList.filter(cartProduct => cartProduct.id !== cartProductId)

            const finalData = [
                ...filteredData,
                editedCartProduct
            ]

            state.dataList = finalData

            // Calculate purchaseDetails

            const subTotal = finalData.reduce(
                (prev, actual) => prev + (actual.salePrice * actual.quantity),
                0
            )
            const taxAmount = subTotal * 0.13
            const total = subTotal + taxAmount

            state.purchaseDetails.subTotal = subTotal
            state.purchaseDetails.taxAmount = taxAmount
            state.purchaseDetails.total = total
        },
        decrementShoppingCartProductQuantity: (state, action) => {
            const cartProductId = action.payload
            const foundCartProduct = current(state).dataList.find(cartProduct => cartProduct.id === cartProductId)


            const newCartProductQuantity =
                foundCartProduct?.quantity === 0
                    ? foundCartProduct?.quantity
                    : foundCartProduct?.quantity - 1

            const editedCartProduct: ShoppingCartProduct = {
                ...foundCartProduct,
                quantity: newCartProductQuantity,
                totalPrice: newCartProductQuantity * foundCartProduct?.salePrice
            }

            const filteredData = current(state).dataList.filter(cartProduct => cartProduct.id !== cartProductId)

            const finalData = [
                ...filteredData,
                editedCartProduct
            ]

            state.dataList = finalData

            // Calculate purchaseDetails

            const subTotal = finalData.reduce(
                (prev, actual) => prev + (actual.salePrice * actual.quantity),
                0
            )
            const taxAmount = subTotal * 0.13
            const total = subTotal + taxAmount

            state.purchaseDetails.subTotal = subTotal
            state.purchaseDetails.taxAmount = taxAmount
            state.purchaseDetails.total = total
        },
        resetShoppingCart: (state, action) => {
            state.dataList = shoppingCartInitialState.dataList
            state.errorText = shoppingCartInitialState.errorText
            state.isLoading = shoppingCartInitialState.isLoading
        },
    },
    extraReducers: () => {

    }
})

// Action creators are generated for each case reducer function
// export const { increment, decrement, incrementByAmount } = counterSlice.actions

export const { addShoppingCartProduct, incrementShoppingCartProductQuantity, decrementShoppingCartProductQuantity } = shoppingCartSlice.actions

export default shoppingCartSlice.reducer