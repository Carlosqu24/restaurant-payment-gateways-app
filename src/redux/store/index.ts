import { configureStore, applyMiddleware, compose } from '@reduxjs/toolkit';
import counterReducer from '../../features/counter/counterSlice';
import thunk, { ThunkAction } from 'redux-thunk';

const middlewareEnhancer = applyMiddleware(thunk)
const composedEnhancers = compose(middlewareEnhancer)

export const store = configureStore({
    reducer: {
        counter: counterReducer,
    },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;