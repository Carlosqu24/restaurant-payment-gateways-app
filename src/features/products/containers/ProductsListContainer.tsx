// @ts-nocheck
import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import ProductListPresentation from "../presentations/ProductListPresentation";

import { fetchAllProducts } from "../../../redux/reducers/productsReducer";
import { addShoppingCartProduct, incrementShoppingCartProductQuantity } from "../../../redux/reducers/shoppingCartReducer";
import { Product, ShoppingCartProduct } from "../../../models/product";

const ProductsListContainer = () => {
  const productsList = useSelector((state: RootState) => state.products.data)
  const dispatch = useDispatch()


  useEffect(() => {
    dispatch(fetchAllProducts())
  }, []);

  const onAddProductToShoppingCart = (product: Product) => {

    const newCartProduct: ShoppingCartProduct = {
      ...product,
      quantity: 1
    }

    dispatch(addShoppingCartProduct(newCartProduct))
  }

  return <ProductListPresentation 
        products={productsList} 
        onAddProductToShoppingCart={onAddProductToShoppingCart} 
      />;
};

export default ProductsListContainer;
