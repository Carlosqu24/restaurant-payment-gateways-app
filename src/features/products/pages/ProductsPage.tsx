import React from "react";
import ProductsListContainer from "../containers/ProductsListContainer";

const Products = () => {
  return (
      <div className="p-4">
        <h2>Productos</h2>

        <ProductsListContainer />
      </div>
  );
};

export default Products;
