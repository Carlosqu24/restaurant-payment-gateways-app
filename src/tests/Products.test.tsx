//@ts-nocheck

import React from "react";
import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";

import {store} from "../redux/store/index"

import ProductsPage from "../features/products/pages/ProductsPage";
import ProductPageObject from "./pageObjects/ProductPageObject";
import { WrappedApp } from "../App";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import productsReducer, { productsReducerTestingState } from "../redux/reducers/productsReducer";
import shoppingCartReducer from "../redux/reducers/shoppingCartReducer";
import ProductsListContainer from "../features/products/containers/ProductsListContainer";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { 
        products: productsReducer,
        shoppingCart: shoppingCartReducer 
      },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

describe("ProductsList Component", () => {
  it("Shoud list the products", () => {
    const {container} = renderWithProviders(
      <ProductsListContainer />,
      {
       preloadedState: {
        products: productsReducerTestingState
       }
      }
    );

    const productsPage = new ProductPageObject(screen);

    // productsPage.hasNotAvailableProductsText()

    // screen.getAllByText("No hay productos disponibles")

    const s = screen.getAllByRole("heading", {value: "As"})

     expect(container).toMatchSnapshot()
    // expect(productsPage.getAddToCartButton()).toBeInTheDocument();
    // expect(productsPage.getAddToCartButton()).not.toBeNull();
  });
});
