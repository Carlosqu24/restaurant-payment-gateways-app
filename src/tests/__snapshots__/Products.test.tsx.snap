// Vitest Snapshot v1

exports[`ProductsList Component > Shoud list the products 1`] = `
<div>
  <ul
    class="grid xl:grid-cols-4 md:grid-cols-2 gap-4"
  >
    <div
      class="
    mt-6 
    flex 
    flex-col 
    self-start 
    rounded-lg 
    shadow-[0_2px_15px_-3px_rgba(0,0,0,0.07),0_10px_20px_-2px_rgba(0,0,0,0.04)] 
    sm:shrink-0 
    sm:grow 
    sm:basis-0 
    cursor-pointer 
    bg-transparent 
  "
    >
      <div
        class="p-6"
        style="background: rgba(117, 117, 117, 0.396); display: flex; align-items: center; justify-content: center;"
      >
        <img
          alt="Hollywood Sign on The Hill"
          class="rounded-t-lg"
          src=""
          style="height: 200px; object-fit: contain;"
        />
      </div>
      <div
        class="py-6 flex justify-between"
      >
        <div>
          <h5
            class="mb-2 text-xl font-medium leading-tight font-extrabold"
          >
            Product 1
          </h5>
          <p
            class="font-extrabold"
          >
            $
            2
          </p>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium cursor-pointer css-i4bv87-MuiSvgIcon-root"
          data-testid="AddShoppingCartIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
          />
        </svg>
      </div>
    </div>
    <div
      class="
    mt-6 
    flex 
    flex-col 
    self-start 
    rounded-lg 
    shadow-[0_2px_15px_-3px_rgba(0,0,0,0.07),0_10px_20px_-2px_rgba(0,0,0,0.04)] 
    sm:shrink-0 
    sm:grow 
    sm:basis-0 
    cursor-pointer 
    bg-transparent 
  "
    >
      <div
        class="p-6"
        style="background: rgba(117, 117, 117, 0.396); display: flex; align-items: center; justify-content: center;"
      >
        <img
          alt="Hollywood Sign on The Hill"
          class="rounded-t-lg"
          src=""
          style="height: 200px; object-fit: contain;"
        />
      </div>
      <div
        class="py-6 flex justify-between"
      >
        <div>
          <h5
            class="mb-2 text-xl font-medium leading-tight font-extrabold"
          >
            Product 2
          </h5>
          <p
            class="font-extrabold"
          >
            $
            6
          </p>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium cursor-pointer css-i4bv87-MuiSvgIcon-root"
          data-testid="AddShoppingCartIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
          />
        </svg>
      </div>
    </div>
  </ul>
</div>
`;
